### DISCUSSION TITLE
Prerender (SSG) with Client Routing behavior

### DISCUSSION THREAD

**Discussion by truongleit** [vikejs/vike#1045] [Discussion] [answered]
Hello. Thanks for an amazing plugin.

I am using your plugin with React to build a landing site. My purpose is to build it as static pages, so it will be entirely data-independent from the CMS. Everything was just fine until I turned on `Client Routing`, app's behavior is no longer as I expected.

For example, after generating static pages, when I navigate from `/home` to `/news` and inspect the Network tab in Chrome's inspector, its behavior is fine (it pulls the data generated in the `index.pageContext.json` file to map it with client's components). However, when it turns to dynamic route (for instance, `/news/@id`), when navigating from a random route to a dynamic route, it still call the hook `onBeforeRender` in `@id.page.tsx` to fetch the data from the CMS, instead of pulling from its `index.pageContext.json` file. 

You can see the process in the attached image, firstly I navigate to `/news`, the app loads the data from the `index.pageContext.json`, but when moving to `/news/21`, it fetchs data from the CMS. Note that, if I refresh the page, it fetchs from the json file.
![Screenshot 2023-08-02 at 00 23 12](https://github.com/brillout/vite-plugin-ssr/assets/30946085/9492c93d-fb05-4476-9eb9-ce9babe57020)

Is there any solution, in which I can use Client Routing but the app only uses generated `json` file from `build` process?

Thank you so much and look forward to your response. @brillout

**Comment by brillout** [comment, marked as answer]
Define `onBeforeRender()` in `.page.server.js` instead of `.page.js`.

**Reply by truongleit** [reply]
It works now. Much appreciation 👍

**Reply by brillout** [reply]
👍 Btw. would your company be up for [sponsoring](https://github.com/sponsors/brillout)?
